File: 03_patterns/swift/mvvm_clean_arc/Shared/Storage/SharedTaskStore.swift
import Foundation
import SwiftData

@Model
final class SharedTaskRecord: Identifiable {
    let id: UUID
    var title: String
    let createdAt: Date

    init(id: UUID = UUID(), title: String, createdAt: Date = .now) {
        self.id = id
        self.title = title
        self.createdAt = createdAt
    }
}

final class SharedTaskStore {
    static let seedPopulatedKey = "SeedPopulated"
    static let defaultTasks: [SharedTaskSeed] = [
        SharedTaskSeed(title: "Plan sprint"),
        SharedTaskSeed(title: "Write unit tests"),
        SharedTaskSeed(title: "Review pull request")
    ]

    private(set) static var shared = SharedTaskStore(
        modelContainer: SharedTaskStore.makeFallbackContainer(),
        userDefaults: .standard
    )

    private let modelContainer: ModelContainer
    private let userDefaults: UserDefaults

    static func configureShared(
        modelContainer: ModelContainer,
        userDefaults: UserDefaults = .standard
    ) {
        shared = SharedTaskStore(
            modelContainer: modelContainer,
            userDefaults: userDefaults
        )
    }

    init(
        modelContainer: ModelContainer,
        userDefaults: UserDefaults = .standard
    ) {
        self.modelContainer = modelContainer
        self.userDefaults = userDefaults
        populateSeedIfNeeded()
    }

    func fetchTasks() -> [SharedTaskRecord] {
        let descriptor = FetchDescriptor<SharedTaskRecord>(
            sortBy: [SortDescriptor(\.createdAt, order: .forward)]
        )

        return (try? modelContext.fetch(descriptor)) ?? []
    }

    func createTask(title: String) -> [SharedTaskRecord] {
        let task = SharedTaskRecord(title: title)
        modelContext.insert(task)
        saveContext()
        return fetchTasks()
    }

    func deleteTask(id: UUID) -> [SharedTaskRecord] {
        let descriptor = FetchDescriptor<SharedTaskRecord>(
            predicate: #Predicate { $0.id == id }
        )

        if let task = try? modelContext.fetch(descriptor).first {
            modelContext.delete(task)
            saveContext()
        }

        return fetchTasks()
    }

    private var modelContext: ModelContext {
        modelContainer.mainContext
    }

    private func populateSeedIfNeeded() {
        guard userDefaults.bool(forKey: Self.seedPopulatedKey) == false else { return }

        for task in Self.defaultTasks {
            modelContext.insert(SharedTaskRecord(title: task.title))
        }

        saveContext()
        userDefaults.set(true, forKey: Self.seedPopulatedKey)
    }

    private func saveContext() {
        do {
            try modelContext.save()
        } catch {
            assertionFailure("Failed to save SharedTaskStore context: \(error)")
        }
    }

    private static func makeFallbackContainer() -> ModelContainer {
        do {
            let schema = Schema([SharedTaskRecord.self])
            let configuration = ModelConfiguration(
                isStoredInMemoryOnly: true
            )
            return try ModelContainer(
                for: schema,
                configurations: [configuration]
            )
        } catch {
            fatalError("Failed to create fallback SharedTaskStore container: \(error)")
        }
    }
}

struct SharedTaskSeed {
    let title: String
}

File: 03_patterns/flutter/lib/shared/storage/shared_task_store.dart
class SharedTaskRecord {
  SharedTaskRecord({
    required this.id,
    required this.title,
    required this.createdAt,
  });

  final String id;
  String title;
  final DateTime createdAt;
}

class SharedTaskSeed {
  const SharedTaskSeed({required this.title});

  final String title;
}

/// Simple injectable flag store, analogous to UserDefaults in the Swift sample.
class SeedFlagStore {
  final Map<String, bool> _values = {};

  bool getBool(String key) => _values[key] ?? false;

  void setBool(String key, bool value) => _values[key] = value;

  void clear() => _values.clear();
}

class SharedTaskStore {
  SharedTaskStore({
    SeedFlagStore? seedFlags,
    List<SharedTaskRecord>? initialRecords,
  }) : _seedFlags = seedFlags ?? SeedFlagStore(),
       // Keep the same list reference when provided so multiple store
       // instances can share persistence (mirrors a shared ModelContainer).
       _records = initialRecords ?? <SharedTaskRecord>[] {
    _populateSeedIfNeeded();
  }

  static const seedPopulatedKey = 'SeedPopulated';
  static const defaultTasks = [
    SharedTaskSeed(title: 'Plan sprint'),
    SharedTaskSeed(title: 'Write unit tests'),
    SharedTaskSeed(title: 'Review pull request'),
  ];

  static SharedTaskStore shared = SharedTaskStore();

  static void configureShared(SharedTaskStore store) {
    shared = store;
  }

  static void resetShared() {
    shared = SharedTaskStore();
  }

  final SeedFlagStore _seedFlags;
  final List<SharedTaskRecord> _records;
  int _idSequence = 0;

  List<SharedTaskRecord> fetchTasks() {
    final sorted = List<SharedTaskRecord>.from(_records)
      ..sort((a, b) => a.createdAt.compareTo(b.createdAt));
    return sorted;
  }

  List<SharedTaskRecord> createTask({required String title}) {
    _records.add(
      SharedTaskRecord(
        id: _nextId(),
        title: title,
        createdAt: DateTime.now(),
      ),
    );
    return fetchTasks();
  }

  List<SharedTaskRecord> deleteTask({required String id}) {
    _records.removeWhere((task) => task.id == id);
    return fetchTasks();
  }

  String _nextId() {
    _idSequence += 1;
    return 'task-$_idSequence-${DateTime.now().microsecondsSinceEpoch}';
  }

  void _populateSeedIfNeeded() {
    if (_seedFlags.getBool(seedPopulatedKey)) {
      return;
    }

    for (final task in defaultTasks) {
      _records.add(
        SharedTaskRecord(
          id: _nextId(),
          title: task.title,
          createdAt: DateTime.now(),
        ),
      );
    }

    _seedFlags.setBool(seedPopulatedKey, true);
  }
}
